### PR TITLE
granular error handling

### PR DESCRIPTION
**Pull request by bdehamer**
#### Summary
Improves the way errors are reported to consumers of this library.

When I went to integrate the new `tuf.getTarget` function in to the npm CLI I realized that we need the ability to distinguish between a TUF target which does not exist and one which exists but failed to download for some other reason -- currently both of these cases raise the same `InternalError` with no way to differentiate them beyond parsing the error message itself.

The change introduced here adds a `code` property to the `InternalError` which can be set to one of a constrained set of values that can be inspected at runtime to determine the specific error condition (this seemed more friendly than proliferating a whole bunch of distinct `Error`-derived classes).

Ultimately, I'd like to extend this pattern to the other errors (`VerificationError`, `ValidationError`, etc) but will leave that for a future change.

### CODE REVIEW

File: src/ca/index.ts [hunk 1]
```diff
@@ -49,7 +49,11 @@ export class CAClient implements CA {
 
       return certificate.signedCertificateEmbeddedSct.chain.certificates;
     } catch (err) {
-      throw new InternalError('error creating signing certificate', err);
+      throw new InternalError({
+        code: 'CREATE_SIGNING_CERTIFICATE_ERROR',
```
Comment (feelepxyz) [inline src/ca/index.ts:53]: Should we have some pattern where we prefix the code with the module/thing?
```suggestion
        code: 'CA_CREATE_SIGNING_CERTIFICATE_ERROR',
```

File: src/tlog/index.ts [hunk 2]
```diff
@@ -92,10 +92,18 @@ export class TLogClient implements TLog {
         try {
           entry = await this.rekor.getEntry(uuid);
         } catch (err) {
-          throw new InternalError('error fetching tlog entry', err);
+          throw new InternalError({
+            code: 'FETCH_TLOG_ENTRY_ERROR',
```
Comment (feelepxyz) [inline src/tlog/index.ts:96]:
```suggestion
            code: 'TLOG_FETCH_ENTRY_ERROR',
```

File: src/tlog/index.ts [hunk 3]
```diff
@@ -92,10 +92,18 @@ export class TLogClient implements TLog {
...
+            message: 'error fetching tlog entry',
+            cause: err,
+          });
         }
       } else {
-        throw new InternalError('error creating tlog entry', err);
+        throw new InternalError({
+          code: 'CREATE_TLOG_ENTRY_ERROR',
```
Comment (feelepxyz) [inline src/tlog/index.ts:103]:
```suggestion
          code: 'TLOG_CREATE_ENTRY_ERROR',
```

File: src/tuf/target.ts [hunk 4]
```diff
@@ -27,7 +27,13 @@ export async function readTarget(
   return new Promise((resolve, reject) => {
     fs.readFile(path, 'utf-8', (err, data) => {
       if (err) {
-        reject(new InternalError(`error reading target: ${err}`));
+        reject(
+          new InternalError({
+            code: 'TUF_TARGET_READ_ERROR',
```
Comment (feelepxyz) [inline src/tuf/target.ts:32]: Thoughts on using some kind of `namespace-prefix_verb/action_resource/thing_error` pattern in codes?
```suggestion
            code: 'TUF_READ_TARGET_ERROR',
```

File: src/tuf/target.ts [hunk 5]
```diff
@@ -43,11 +49,18 @@ async function getTargetPath(tuf: Updater, target: string): Promise<string> {
   try {
     targetInfo = await tuf.refresh().then(() => tuf.getTargetInfo(target));
   } catch (err) {
-    throw new InternalError(`error refreshing TUF metadata: ${err}`);
+    throw new InternalError({
+      code: 'TUF_METADATA_REFRESH_ERROR',
```
Comment (feelepxyz) [inline src/tuf/target.ts:53]:
```suggestion
      code: 'TUF_REFRESH_METADATA_ERROR',
```

File: src/tuf/target.ts [hunk 6]
```diff
@@ -58,7 +71,11 @@ async function getTargetPath(tuf: Updater, target: string): Promise<string> {
     try {
       path = await tuf.downloadTarget(targetInfo);
     } catch (err) {
-      throw new InternalError(`error downloading target: ${err}`);
+      throw new InternalError({
+        code: 'TUF_TARGET_DOWNLOAD_ERROR',
```
Comment (feelepxyz) [inline src/tuf/target.ts:75]:
```suggestion
        code: 'TUF_DOWNLOAD_TARGET_ERROR',
```

File: src/tuf/target.ts [hunk 7]
```diff
@@ -43,11 +49,18 @@ async function getTargetPath(tuf: Updater, target: string): Promise<string> {
...
+      code: 'TUF_METADATA_REFRESH_ERROR',
+      message: 'error refreshing TUF metadata',
+      cause: err,
+    });
   }
 
   if (!targetInfo) {
-    throw new InternalError(`target ${target} not found`);
+    throw new InternalError({
+      code: 'TUF_TARGET_NOT_FOUND',
```
Comment (feelepxyz) [inline src/tuf/target.ts:61]: maybe a bit funky to fit the pattern?
```suggestion
      code: 'TUF_FIND_TARGET_ERROR',
```

**Comment by bdehamer** [comment]
@feelepxyz I like your proposed codes. Updated.

**Review by feelepxyz — APPROVED** [review]
👍

**Pull request merged**